ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,436 @@
+		bc_convert_to_vector(n1_vector + n1_vector_count, n1, n1_readable_len - n1_read);
+	}
+	bc_convert_to_vector(n2_vector, n2, n2_len);
+
+	/* Do the divide */
+	if (n2_arr_size == 1) {
+		bc_fast_div(n1_vector, n1_arr_size, n2_vector[0], quot_vector, quot_arr_size);
+	} else {
+		bc_standard_div(n1_vector, n1_arr_size, n2_vector, n2_arr_size, n2_len, quot_vector, quot_arr_size);
+	}
+
+	/* Convert to bc_num */
+	char *qptr = (*quot)->n_value;
+	char *qend = qptr + quot_len - 1;
+
+	for (i = 0; i < quot_real_arr_size - 1; i++) {
+#if BC_VECTOR_SIZE == 4
+		bc_write_bcd_representation(quot_vector[i], qend - 3);
+		qend -= 4;
+#else
+		bc_write_bcd_representation(quot_vector[i] / 10000, qend - 7);
+		bc_write_bcd_representation(quot_vector[i] % 10000, qend - 3);
+		qend -= 8;
+#endif
+	}
 
-/* The full division routine. This computes N1 / N2.  It returns
-   true if the division is ok and the result is in QUOT.  The number of
-   digits after the decimal point is SCALE. It returns false if division
-   by zero is tried. The algorithm is found in Knuth Vol 2. p237. */
-bool bc_divide(bc_num n1, bc_num n2, bc_num *quot, int scale)
+	while (qend >= qptr) {
+		*qend-- = quot_vector[i] % BASE;
+		quot_vector[i] /= BASE;
+	}
+
+	efree(n1_vector);
+}
+
+bool bc_divide(bc_num n1, bc_num n2, bc_num *quot, size_t scale)
 {
-	bc_num qval;
-	unsigned char *num1, *num2;
-	unsigned char *ptr1, *ptr2, *n2ptr, *qptr;
-	int scale1, val;
-	unsigned int len1, len2, scale2, qdigits, extra, count;
-	unsigned int qdig, qguess, borrow, carry;
-	unsigned char *mval;
-	unsigned int norm;
-	bool zero;
-
-	/* Test for divide by zero. */
+	/* divide by zero */
 	if (bc_is_zero(n2)) {
 		return false;
 	}
 
-	/* Test for divide by 1. If it is we must truncate. */
-	if (n2->n_scale == 0 && n2->n_len == 1 && *n2->n_value == 1) {
-		qval = bc_new_num (n1->n_len, scale);
-		qval->n_sign = (n1->n_sign == n2->n_sign ? PLUS : MINUS);
-		memcpy(qval->n_value, n1->n_value, n1->n_len + MIN(n1->n_scale, scale));
-		bc_free_num (quot);
-		*quot = qval;
+	bc_free_num(quot);
+
+	/* If n1 is zero, the quotient is always zero. */
+	if (bc_is_zero(n1)) {
+		*quot = bc_copy_num(BCG(_zero_));
+		return true;
 	}
 
-	/* Set up the divide.  Move the decimal point on n1 by n2's scale.
-	   Remember, zeros on the end of num2 are wasted effort for dividing. */
-	scale2 = n2->n_scale;
-	n2ptr = (unsigned char *) n2->n_value + n2->n_len + scale2 - 1;
-	while ((scale2 > 0) && (*n2ptr == 0)) {
-		scale2--;
-		n2ptr--;
+	char *n1ptr = n1->n_value;
+	char *n1end = n1ptr + n1->n_len + n1->n_scale - 1;
+	size_t n1_len = n1->n_len;
+	size_t n1_scale = n1->n_scale;
+
+	char *n2ptr = n2->n_value;
+	char *n2end = n2ptr + n2->n_len + n2->n_scale - 1;
+	size_t n2_len = n2->n_len;
+	size_t n2_scale = n2->n_scale;
+	size_t n2_int_right_zeros = 0;
+
+	/* remove n2 trailing zeros */
+	while (*n2end == 0 && n2_scale > 0) {
+		n2end--;
+		n2_scale--;
+	}
+	while (*n2end == 0) {
+		n2end--;
+		n2_int_right_zeros++;
 	}
```

Comment:
This feels like it's removing trailing nul bytes and not `0`?

Reply:
Do you mean `'0'` and not `0`?

If so, then 0 in bc_num is the same value as a nul byte, so this is correct.

Reply:
Okay I forgot that detail